lib/markbind/lib/parser.js

```diff
@@ -91,6 +92,7 @@ Parser.prototype._preprocess = function (node, context, config) {
   element.attribs = element.attribs || {};
   element.attribs[ATTRIB_CWF] = path.resolve(context.cwf);
   if (element.name === 'include') {
+    const isBoilerplate = _.hasIn(element.attribs, 'boilerplate');
```

Comment:
The check `_.hasIn(element.attribs, 'boilerplate');` appears 3 times.

There are 5 switch-cases based on `element` type:
- `if (element.name === 'include')`
- `else if (element.name === 'dynamic-panel')`
- `else if ((element.name === 'morph' || element.name === 'panel') && _.hasIn(element.attribs, 'src'))`
- `else if (element.children && element.children.length > 0)`
- default

The check is done in 3 of 5 cases, and doesn't hurt to do for the other 2.

I propose assigning `isBoilerplate` and `element.attribs[ATTRIB_BOILERPLATE]` before the first `if`.
Then we avoid `pathWithinBoilerplate || path.basename(filePath)` in `calculateBoilerplateFilePath`.

Reply:
In fact, this is not the only repetition. See `isUrl`, `includeSrc`, `includeSrcPath` etc.

I've thought about the other repetitions before, but opted for duplication "as needed" previously because I didn't have time to explore the implications. However, with my current understanding, this seems like a trade-off between DRY and unnecessary/premature computation.

Alternatively, those could be abstracted into helper functions.

@Gisonrg Your thoughts?

Reply:
@acjh 
Something like this?
https://github.com/MarkBind/markbind/pull/167/commits/6173f3c424c336ac7a9452004e4a4e8d354aa78a

Reply:
@rachx Nice cleanup :)

> I propose assigning `isBoilerplate` and `element.attribs[ATTRIB_BOILERPLATE]` before the first `if`.
Then we avoid `pathWithinBoilerplate || path.basename(filePath)` in `calculateBoilerplateFilePath`.

This was not addressed? Maybe even
`calculateBoilerplateFilePath(element.attribs.boilerplate || path.basename(filePath), config);` might be clearer.